### ISSUE TITLE
Umka fails on SPARC due to misaligned pointers 

### ISSUE THREAD

**Issue by tankf33der** [vtereshkov/umka-lang#20]
Solaris 11, clang, SPARC.

кое что работает:
```
$ ../umka gc.um 
See any memory leak warnings?
$
```

но все остальное прочно сломано
```
$ ../umka tour.um 
Bus Error (core dumped)
```

надо чинить все предыдущие и другая платформа сама заработает.

**Comment by vtereshkov** [comment]
I guess SPARC disallows misaligned addresses. OK, I'll think of it.

**Comment by vtereshkov** [comment]
Did you try the `-xmemalign` compiler option?

**Comment by tankf33der** [comment]
> 
> 
> Did you try the `-xmemalign` compiler option?

No. How could you assume I have Sun Studio? I would like to stay on gcc-clang side as I wrote before.

**Comment by tankf33der** [comment]
```
Core was generated by `./umka tests/tour.um'.
Program terminated with signal SIGBUS, Bus error.
#0  0x000000010000b994 in constAssign (consts=0x10012ad38, lhs=0x10012b359, rhs=0xffffffff7fffd9b0, typeKind=TYPE_REAL, size=8) at src/umka_const.c:39
39              case TYPE_REAL:         *(double   *)lhs = rhs->realVal;        break;
[Current thread is 4 (Thread 1 (LWP 1))]
(gdb) bt
#0  0x000000010000b994 in constAssign (consts=0x10012ad38, lhs=0x10012b359, rhs=0xffffffff7fffd9b0, typeKind=TYPE_REAL, size=8) at src/umka_const.c:39
#1  0x0000000100013ecc in parseArrayOrStructLiteral (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:987
#2  0x0000000100014258 in parseCompositeLiteral (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1039
#3  0x00000001000143a0 in parseTypeCastOrCompositeLiteral (comp=0x100129900, ident=0x0, type=0xffffffff7fffe468, constant=0xffffffff7fffe460, isVar=0xffffffff7fffdccb, 
    isCall=0xffffffff7fffdcca) at src/umka_expr.c:1054
#4  0x00000001000154ec in parseDesignator (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460, isVar=0xffffffff7fffdccb, isCall=0xffffffff7fffdcca)
    at src/umka_expr.c:1270
#5  0x00000001000155b8 in parseFactor (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1293
#6  0x0000000100015d1c in parseTerm (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1417
#7  0x0000000100015ec0 in parseRelationTerm (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1441
#8  0x0000000100016024 in parseRelation (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1465
#9  0x00000001000161d8 in parseLogicalTerm (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1491
#10 0x0000000100016398 in parseExpr (comp=0x100129900, type=0xffffffff7fffe468, constant=0xffffffff7fffe460) at src/umka_expr.c:1529
#11 0x0000000100013e20 in parseArrayOrStructLiteral (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:981
#12 0x0000000100014258 in parseCompositeLiteral (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1039
#13 0x00000001000143a0 in parseTypeCastOrCompositeLiteral (comp=0x100129900, ident=0x100a5b0f0, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0, 
    isVar=0xffffffff7fffe77b, isCall=0xffffffff7fffe77a) at src/umka_expr.c:1054
#14 0x00000001000154ec in parseDesignator (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0, isVar=0xffffffff7fffe77b, isCall=0xffffffff7fffe77a)
    at src/umka_expr.c:1270
#15 0x00000001000155b8 in parseFactor (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1293
#16 0x0000000100015d1c in parseTerm (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1417
#17 0x0000000100015ec0 in parseRelationTerm (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1441
#18 0x0000000100016024 in parseRelation (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1465
#19 0x00000001000161d8 in parseLogicalTerm (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1491
#20 0x0000000100016398 in parseExpr (comp=0x100129900, type=0xffffffff7fffecc8, constant=0xffffffff7fffecc0) at src/umka_expr.c:1529
#21 0x000000010001c8dc in parseDeclAssignmentStmt (comp=0x100129900, name=0xffffffff7fffed98 "q", constExpr=true, exported=false) at src/umka_stmt.c:131
#22 0x000000010000e9fc in parseShortVarDecl (comp=0x100129900) at src/umka_decl.c:498
#23 0x000000010000ee08 in parseDecl (comp=0x100129900) at src/umka_decl.c:549
#24 0x000000010000eee0 in parseDecls (comp=0x100129900) at src/umka_decl.c:565
#25 0x000000010000f384 in parseModule (comp=0x100129900) at src/umka_decl.c:636
#26 0x000000010000f408 in parseProgram (comp=0x100129900) at src/umka_decl.c:648
#27 0x000000010000b300 in compilerCompile (comp=0x100129900) at src/umka_compiler.c:164
#28 0x0000000100008740 in umkaCompile (umka=0x100129900) at src/umka_api.c:81
#29 0x000000010000816c in main (argc=2, argv=0xffffffff7ffff8a8) at src/umka.c:83
```

**Comment by vtereshkov** [comment]
See also #19

**Comment by tankf33der** [comment]
still crashes, the same back trace above.

**Comment by vtereshkov** [comment]
Yes, I have fixed only the smaller problem with signed integer overflow UB, which was also in the sanitizer report. The misaligned pointer problem is still there.

**Comment by vtereshkov** [comment]
@tankf33der This problem is expected to be indirectly fixed now (see #19). Please close the issue is the crash no longer occurs.

**Comment by vtereshkov** [comment]
Successfully tested on QEMU, 32-bit SPARC, OpenBSD 5.2, GCC 4.2.1. No bus errors detected. Closing.
